sdk/src/wallet/migration/chrysalis.rs

```diff
@@ -0,0 +1,310 @@
+) -> Result<()> {
+    let storage_path_string = storage_path.into();
+    // `/db` will be appended to the chrysalis storage path, because that's how it was done in the chrysalis wallet
+    let chrysalis_storage_path = &(*storage_path_string).join("db");
+
+    let chrysalis_data = get_chrysalis_data(chrysalis_storage_path, password)?;
+
+    // create new accounts base on previous data
+    let (new_accounts, secret_manager_dto) = migrate_from_chrysalis_data(&chrysalis_data, &storage_path_string, false)?;
+
+    // convert to string keys
+    let chrysalis_data_with_string_keys = chrysalis_data
+        .into_iter()
+        .map(|(k, v)| {
+            Ok((
+                String::from_utf8(k).map_err(|_| Error::Migration("invalid utf8".into()))?,
+                v,
+            ))
+        })
+        .collect::<Result<HashMap<String, String>>>()?;
+
+    log::debug!(
+        "Chrysalis data: {}",
+        serde_json::to_string_pretty(&chrysalis_data_with_string_keys)?
+    );
+
+    let mut opts = Options::default();
+    opts.create_if_missing(true);
+    opts.create_missing_column_families(true);
+    let stardust_db = DB::open(&opts, storage_path_string).unwrap();
+
+    // store chrysalis data in a new key
+    stardust_db.put(
+        CHRYSALIS_STORAGE_KEY,
+        serde_json::to_string(&chrysalis_data_with_string_keys)?,
+    )?;
+    // write new accounts to db (with account indexation)
+    let accounts_indexation_data: Vec<u32> = new_accounts.iter().map(|account| account.index).collect();
+    stardust_db.put(
+        ACCOUNTS_INDEXATION_KEY,
+        serde_json::to_string(&accounts_indexation_data)?,
+    )?;
+    for new_account in new_accounts {
+        stardust_db.put(
+            &format!("{ACCOUNT_INDEXATION_KEY}{}", new_account.index),
+            serde_json::to_string(&new_account)?,
+        )?;
+    }
+
+    if let Some(secret_manager_dto) = secret_manager_dto {
+        // This is required for the secret manager to be loaded
+        stardust_db.put(WALLET_INDEXATION_KEY, format!("{{ \"coinType\": {IOTA_COIN_TYPE}}}"))?;
+        stardust_db.put(SECRET_MANAGER_KEY, secret_manager_dto)?;
+    }
+
+    // set db migration version
+    let migration_version = crate::wallet::migration::migrate_4::Migrate::version();
+    stardust_db.put(MIGRATION_VERSION_KEY, serde_json::to_string(&migration_version)?)?;
+
+    drop(stardust_db);
+
+    // remove old db
+    std::fs::remove_dir_all(chrysalis_storage_path)?;
+
+    Ok(())
+}
+
+pub(crate) fn migrate_from_chrysalis_data(
+    chrysalis_data: &HashMap<Vec<u8>, String>,
+    storage_path: &Path,
+    // in stronghold the keys are hashed first
+    stronghold: bool,
+) -> Result<(Vec<AccountDetailsDto>, Option<String>)> {
+    let mut new_accounts: Vec<AccountDetailsDto> = Vec::new();
+    let mut secret_manager_dto: Option<String> = None;
+
+    let account_indexation_key = if stronghold {
+        key_to_chrysalis_key(b"iota-wallet-account-indexation")
```

Comment:
I would have this function to take the bool, less outside ifs

Reply:
Done